proposals/0000-revise-partition-method.md

```diff
@@ -0,0 +1,104 @@
+* Authors: [Lorenzo Racca](https://github.com/lorenzoracca), [Jeff Hajewski](https://github.com/j-haj), [Nate Cook](https://github.com/natecook1000)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+This proposal revises the signature for the collection partition algorithm. Partitioning is a foundational API for sorting and for searching through sorted collections.
+
+- Swift-evolution thread: [Feedback from standard library team](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20160502/016729.html)
+- Swift Bug: [SR-1965](https://bugs.swift.org/browse/SR-1965)
+
+## Motivation
+
+Based on feedback during the review of proposal [SE-0074, Implementation of Binary Search Functions][se-74] and the [list of open issues affecting standard library API stability][list], this is a revised proposal focused only on the existing collection partition method.
+
+The standard library's current `partition` methods, which partition a mutable collection using a binary predicate based on the value of the first element of a collection, are used by the standard library's sorting algorithm but don't offer more general partitioning functionality. A more general partition algorithm using a unary (single-argument) predicate would be more flexible and generally useful.
+
+[se-74]: https://github.com/apple/swift-evolution/blob/master/proposals/0074-binary-search.md
+[list]: https://gist.github.com/gribozavr/37e811f12b27c6365fc88e6f9645634d
+
+## Proposed solution
+
+The standard library should replace the two existing `partition` methods with a single method taking a unary predicate called `partition(by:)`. `partition(by:)` rearranges the elements of the collection according to the predicate, such that after partitioning there is a pivot index `p` where no element before `p` satisfies the predicate and every element at and after `p` *does* satisfy the predicate.
+
+```swift
+var n = [30, 40, 20, 30, 30, 60, 10]
+let p = n.partition(by: { $0 > 30 })
+// n == [30, 10, 20, 30, 30, 60, 40]
+// p == 5
+```
+
+After partitioning is complete, the predicate returns `false` for every element in `n.prefix(upTo: p)` and `true` for every element in `n.suffix(from: p)`.
+
+## Detailed design
+
+`partition(by:)` should be added as a `MutableCollection` requirement with default implementations for mutable and bidirectional mutable collections. Any mutable collection can be partitioned, but the bidirectional algorithm generally performs far fewer assignments.
+
+The proposed APIs are collected here:
+
+```swift
+protocol MutableCollection {
+    // existing requirements
+    
+    /// Reorders the elements of the collection such that all the
+    /// elements that match the predicate are ordered after all the
+    /// elements that do not match the predicate.
+    ///
+    /// - Parameter belongsInSecondPartition: A predicate that returns 
+    ///   `true` if the argument belongs in the second partition; 
+    ///   otherwise, `false`.
```

Comment:
The parameter name and its documentation are circular.  To break that circle, you could choose a name that refers to terms in the summary rather than “second partition”, or you could describe the parameter with different words.